server_conn.go

```diff
@@ -331,8 +337,12 @@ func (c *serverConn) setUpgrading(name string, s transport.Server) {
 }
 
 func (c *serverConn) upgraded() {
-	c.transportLocker.Lock()
+	if c.upgrading == nil {
+		fmt.Println("Bypass double upgrade from IP", c.request.RemoteAddr)
```

Comment:
It all looks great, the only comment I have is to remove this piece of logging here. This library doesn't really log anything at the moment so this feels pretty out of place. Either remove it or use `log.Println` (which I'm not super happy about because the user can't disable that. It would be nice if this library used a configurable logger for these cases). But I don't really see the point in logging the IP address now that the vulnerability has been fixed.

Reply:
fair call - removed.

Was logging the IP so - if we get shot at, we can see where the bullet came from as well as deflecting it.

Bit OTT, and yes, it probably is out of place in here.